File: web/packages/core/src/backoff.ts
/**
 * Flap-resistant reconnect backoff — pure functions (fleet-daemon lesson,
 * docs/11). spec: docs/08-protocol.md#reconnection
 */

export interface BackoffPolicy {
  baseMs: number;
  factor: number;
  capMs: number;
  /** ±fraction of the delay. */
  jitter: number;
  /** Attempts reset only after this long connected. */
  stableResetMs: number;
}

export const DEFAULT_BACKOFF: BackoffPolicy = {
  baseMs: 500,
  factor: 2,
  capMs: 30_000,
  jitter: 0.2,
  stableResetMs: 30_000,
};

/** Delay for `attempt` (0-based). `random` ∈ [0,1) for deterministic tests. */
export function backoffDelay(attempt: number, policy: BackoffPolicy = DEFAULT_BACKOFF, random: () => number = Math.random): number {
  const raw = Math.min(policy.capMs, policy.baseMs * Math.pow(policy.factor, Math.max(0, attempt)));
  const spread = raw * policy.jitter;
  return Math.round(raw - spread + random() * 2 * spread);
}

export class Backoff {
  private attempt = 0;
  private connectedAt: number | null = null;
  constructor(
    private readonly policy: BackoffPolicy = DEFAULT_BACKOFF,
    private readonly random: () => number = Math.random,
  ) {}

  get attempts(): number {
    return this.attempt;
  }

  /** Delay before the next attempt; increments the attempt counter. */
  next(): number {
    return backoffDelay(this.attempt++, this.policy, this.random);
  }

  markConnected(now: number): void {
    this.connectedAt = now;
  }

  /** On disconnect: attempts reset only if the connection was stable. */
  markDisconnected(now: number): void {
    if (this.connectedAt !== null && now - this.connectedAt >= this.policy.stableResetMs) {
      this.attempt = 0;
    }
    this.connectedAt = null;
  }

  reset(): void {
    this.attempt = 0;
    this.connectedAt = null;
  }
}

File: agent/src/core/backoff.cpp
#include "backoff.hpp"

#include <algorithm>
#include <cmath>
#include <random>

namespace fjarr {

Backoff::Backoff(BackoffPolicy policy, std::function<double()> random) : policy_(policy), random_(std::move(random)) {
    if (!random_) {
        random_ = [] {
            static thread_local std::mt19937 gen{std::random_device{}()};
            return std::uniform_real_distribution<double>(0.0, 1.0)(gen);
        };
    }
}

std::chrono::milliseconds Backoff::next() {
    const double base = static_cast<double>(policy_.initial.count()) * std::pow(policy_.factor, attempts_);
    const double capped = std::min(base, static_cast<double>(policy_.max.count()));
    const double jitter = 1.0 + (random_() * 2.0 - 1.0) * policy_.jitter;
    attempts_++;
    return std::chrono::milliseconds(static_cast<long>(std::max(0.0, capped * jitter)));
}

void Backoff::mark_connected(std::chrono::steady_clock::time_point now) {
    connected_ = true;
    connected_at_ = now;
}

void Backoff::mark_disconnected(std::chrono::steady_clock::time_point now) {
    if (connected_ && now - connected_at_ >= policy_.stable_reset) attempts_ = 0;
    connected_ = false;
}

void Backoff::reset() {
    attempts_ = 0;
    connected_ = false;
}

} // namespace fjarr
